File: format.py
#!/usr/bin/env python3
import os
import subprocess
import sys

# Directories to search
directories = ["arch", "include", "kern", "lib", "user"]

# File extensions to process
extensions = (".c", ".h")

def gather_files():
    files_to_format = []
    for directory in directories:
        if not os.path.isdir(directory):
            print(f"Warning: Directory '{directory}' not found, skipping.", file=sys.stderr)
            continue
        for root, _, files in os.walk(directory):
            for filename in files:
                if filename.endswith(extensions):
                    full_path = os.path.join(root, filename)
                    files_to_format.append(full_path)
    return files_to_format

def main():
    files = gather_files()
    if not files:
        print("No files found to format.")
        sys.exit(0)

    # Build the clang-format command with the -i flag and all file paths.
    cmd = ["clang-format", "-i"] + files

    try:
        subprocess.run(cmd, check=True)
        print("clang-format executed successfully.")
    except subprocess.CalledProcessError as e:
        print(f"Error: clang-format returned non-zero exit status: {e}", file=sys.stderr)
        sys.exit(1)
    except FileNotFoundError:
        print("Error: clang-format not found. Is it installed and in your PATH?", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()

File: format.rb
#!/usr/bin/env ruby
# Code formatter - runs clang-format on all C/header files

require 'optparse'
require 'find'

# Default configuration
options = {
  dirs: %w[arch include kern lib user],
  clang_format: 'clang-format',
  check: false,
  verbose: false
}

parser = OptionParser.new do |opts|
  opts.banner = "Usage: #{$0} [options]"
  opts.separator ""
  opts.separator "Format all C and header files using clang-format"
  opts.separator ""

  opts.on("-c", "--check", "Check formatting without modifying files") do
    options[:check] = true
  end

  opts.on("-v", "--verbose", "Show files being formatted") do
    options[:verbose] = true
  end

  opts.on("--clang-format PATH", "Path to clang-format binary") do |v|
    options[:clang_format] = v
  end

  opts.on("-h", "--help", "Show this help message") do
    puts opts
    exit
  end
end

parser.parse!

# Find all C and header files
files = []
options[:dirs].each do |dir|
  next unless Dir.exist?(dir)

  Find.find(dir) do |path|
    next unless File.file?(path)
    next unless path.end_with?('.c', '.h')
    files << path
  end
end

if files.empty?
  warn "No C or header files found in: #{options[:dirs].join(', ')}"
  exit 1
end

puts "Formatting #{files.size} file(s)..." if options[:verbose]

# Build clang-format command
cmd = [options[:clang_format]]

if options[:check]
  # Check mode: dry-run to verify formatting
  cmd << '--dry-run'
  cmd << '-Werror'
end

cmd << '-i'  # In-place formatting
cmd += files

# Show files being formatted in verbose mode
if options[:verbose]
  files.each { |f| puts "  #{f}" }
end

# Execute clang-format
system(*cmd) or exit($?.exitstatus)

puts "Done!" if options[:verbose] && !options[:check]
puts "All files are properly formatted." if options[:check] && $?.success?
